### ISSUE TITLE
Why does zipping with periodic() hangs?

### ISSUE THREAD

**Issue by iwinux** [cujojs/most#466]
### Summary

I want to print a stream of increasing numbers every 1s, but the code below hangs Node.js forever:

```javascript
// test.js
const { iterate, periodic } = require('most')

iterate(x => x + 1, 0)
  .take(10)
  .zip(x => x, periodic(1000))
  .forEach(x => console.log(x))
```

while this

```javascript
iterate(x => x + 1, 0)
  .take(10)
  .forEach(x => console.log(x))
```

terminates as expected.

### Expected result

Expected output would be:

```sh
$ node test.js
0
1 # after 1s
2 # after 2s
...
```

### Actual Result

```sh
$ node test.js
0 # <- hangs here and CPU usage goes up
```

### Versions

- most.js: 1.5.0

**Comment by briancavalier** [comment]
Hi @iwinux, thanks for reporting this.  The sample code to reproduce was very helpful!  It appears to be a regression in `take` introduced in #428 (technically, a regression in `slice`, on which `take` is based).  The regression is causing iterate to run forever, and continuously preempt the periodic task.

The fix is relatively simple, and I should have time tonight to create a branch so you can try it.

If you're indeed blocked by needing a periodic counter, then this will work and hopefully unblock you until the fix is ready:

```js
periodic(1000)
  .slice(1, 10)
  .scan((x, _) => x + 1, 0)
  .forEach(x => console.log(x))
```

I'm guessing your example is a simplified version of a real use case, so this may not help, but figured it was worth a shot.  Cheers!

**Comment by briancavalier** [comment]
@iwinux There's a candidate fix in #467 ([fix-466 branch](https://github.com/cujojs/most/tree/fix-466)).  Could you give it a try and let us know if it works for you?

**Comment by iwinux** [comment]
```sh
$ git clone https://github.com/cujojs/most.git
$ git checkout fix-466-2
$ npm install # some build error coming from contextify
$ npm run build
```

This version seems to fix it. The test case no longer hangs. Thanks!

BTW: should `iterate` always be combined with `take` or `slice`? How to create a lazy & infinite stream that emits one value every second?

**Comment by briancavalier** [comment]
> This version seems to fix it.

:+1: Thanks for trying it.  We'll get this merged and release a patch version soon.

> should iterate always be combined with take or slice?

It depends on your use case.  Usually, yes, because `iterate`, by itself, is always infinite, as are `unfold` and `periodic`, and any stream created by applying `from` to any existing infinite source, such as an infinite iterable.

However, the step function you provide to iterate [may return a promise](https://github.com/cujojs/most/blob/master/docs/api.md#mostiterate)--i.e. instead of returning an `number`, it can return `Promise<number>`.  Thus, the promise can be used to control the timing.

> How to create a lazy & infinite stream that emits one value every second?

It depends on what "value" you want the events to have.  Usually, [`periodic`](https://github.com/cujojs/most/blob/master/docs/api.md#mostperiodic) is the right place to start.  For example, if you want an infinite stream containing `1` every second:

```js
periodic(1000).constant(1)
```

If you want an infinite stream that starts at `0` and increases by `n` every 500 milliseconds:

```js
const byN = n =>
  periodic(500).skip(1).constant(n).scan((x, n) => x + n, 0)
```

**Comment by briancavalier** [comment]
@iwinux The fix was released in [most 1.5.1](https://github.com/cujojs/most/releases/tag/1.5.1).  Thanks again!
